base/array.jl

```diff
@@ -2493,13 +2491,13 @@ function _findall(f::Function, I::Vector, A::AbstractVector{Bool})
     len = length(I)
     while cnt ≤ len
         @inbounds I[cnt] = i
-        cnt += f(@inbounds A[i])
+        cnt += f(A[i])
         i = nextind(A, i)
     end
-    cnt - 1 == len ? I : resize!(I, cnt - 1)
+    I
 end
 
-findall(f::Function, A::AbstractArray{Bool}) = _findall(f, A)
+findall(f::Union{typeof(!), typeof(identity)}, A::AbstractArray{Bool}) = _findall(f, A)
```

Comment:
@aviatesk this seems like another case where ideally we should be dispatching on the `Core.Compiler.infer_effect` result for `f(A)` to the best choice of algorithm.